File: functions/src/lib/name_key.ts
/**
 * Item name normalization and key derivation for
 * `groups/{groupId}/purchaseHistorySummaries/{nameKey}`.
 *
 * Spec: docs/ドラフト/AI提案機能/01-履歴データ設計.md §4
 *
 * Algorithm (MUST stay in sync with any future Dart-side implementation,
 * e.g. Issue #43 「よく買う物リスト」):
 *   1. Unicode NFKC normalization (full-width alnum -> half-width,
 *      half-width kana -> full-width kana, etc.)
 *   2. Trim leading/trailing whitespace and collapse runs of internal
 *      whitespace into a single space.
 *   3. Lowercase (`toLowerCase()`).
 *   4. Hash the result with SHA-1 and encode as base64url to obtain the
 *      Firestore document ID (avoids `/` and length issues with raw
 *      product names). The raw display text is kept separately in the
 *      `name` field — this function only derives the key.
 *
 * This module has no dependency on `firebase-functions` / `firebase-admin`
 * so it can be unit-tested directly and ported to other runtimes.
 */
import { createHash } from "crypto";

/**
 * Normalize a raw item name for use as a summary grouping key.
 *
 * See module doc comment for the exact algorithm steps.
 */
export function normalizeName(raw: string): string {
  return raw
    .normalize("NFKC")
    .trim()
    .replace(/\s+/g, " ")
    .toLowerCase();
}

/**
 * Derive the `purchaseHistorySummaries` document ID for a raw item name.
 *
 * `base64url(SHA-1(normalizeName(raw)))`.
 */
export function nameKeyOf(raw: string): string {
  const normalized = normalizeName(raw);
  return createHash("sha1").update(normalized, "utf8").digest("base64url");
}

File: lib/core/utils/name_key.dart
// アイテム名の正規化ユーティリティ。
// `functions/src/lib/name_key.ts` の `normalizeName` と**同一のアルゴリズム**を実装する。
// 重複判定（よく買う物テンプレートとアクティブアイテム名との突合）に使う。
//
// 両実装の出力一致は `test/fixtures/name_normalization_cases.json` を
// Dart / TypeScript 双方のテストが読むことで担保している（#88）。
// アルゴリズムを変更する場合は必ず両方の実装とフィクスチャを同時に更新すること。
//
// TS 側にある `nameKeyOf`（`base64url(SHA-1(normalizeName(raw)))`）は Dart へ移植していない。
// `nameKey` は `groups/{groupId}/purchaseHistorySummaries/{nameKey}` のドキュメント ID 専用で、
// クライアントはこのコレクションを読まない（購入履歴画面 #42 が読むのは `itemHistory`）ため。
// クライアントからサマリーを直接引く必要が生じた時点で移植する。

import 'package:unorm_dart/unorm_dart.dart' as unorm;

/// アイテム名を正規化する（重複判定用）。
///
/// 処理内容（`functions/src/lib/name_key.ts` の `normalizeName` と同順・同内容）:
///   1. Unicode NFKC 正規化（全角英数→半角、半角カナ→全角カナ、全角スペース→半角スペース等）
///   2. 前後の空白を除去（`trim`）
///   3. 連続する空白を 1 つに圧縮（`\s` は ECMAScript 準拠で全角スペース等も含む）
///   4. 小文字化（`toLowerCase`）
///
/// Dart の標準ライブラリに Unicode 正規化が無いため NFKC は `unorm_dart` で行う
/// （純 Dart 実装のため Web / Android / iOS すべてで動作する）。
String normalizeName(String raw) {
  return unorm.nfkc(raw).trim().replaceAll(RegExp(r'\s+'), ' ').toLowerCase();
}
